File: lib/admin/features/tickets_jobs/services/admin_ticket_job_service.dart
import 'package:cloud_firestore/cloud_firestore.dart' as fs;

import '../../../../core/models/enums.dart';
import '../../../../core/services/firebase_service.dart';
import '../../../../features/customer_ticket/models/ticket_model.dart';
import '../../../../features/service_lifecycle/models/job_model.dart';
import '../../../core/admin_audit_service.dart';

class TicketPage {
  final List<Ticket> tickets;
  final fs.DocumentSnapshot<Map<String, dynamic>>? lastDoc;
  final bool hasMore;
  const TicketPage({required this.tickets, required this.lastDoc, required this.hasMore});
}

class JobPage {
  final List<Job> jobs;
  final fs.DocumentSnapshot<Map<String, dynamic>>? lastDoc;
  final bool hasMore;
  const JobPage({required this.jobs, required this.lastDoc, required this.hasMore});
}

/// Cross-user oversight of the marketplace's two core records. Reads and
/// writes here work because `firestore.rules` grants `isAdmin()`
/// unrestricted access on both `tickets` and `jobs` — the same admin bypass
/// that lets the mobile app's per-user scoping stay tight for everyone else.
class AdminTicketJobService {
  AdminTicketJobService._();
  static final AdminTicketJobService instance = AdminTicketJobService._();

  fs.CollectionReference<Map<String, dynamic>> get _ticketsCol =>
      FirebaseService.instance.firestore.collection('tickets');
  fs.CollectionReference<Map<String, dynamic>> get _openTicketsCol =>
      FirebaseService.instance.firestore.collection('openTickets');
  fs.CollectionReference<Map<String, dynamic>> get _jobsCol =>
      FirebaseService.instance.firestore.collection('jobs');

  Future<TicketPage> fetchTicketPage({
    TicketStatus? status,
    fs.DocumentSnapshot<Map<String, dynamic>>? startAfter,
    int limit = 25,
  }) async {
    fs.Query<Map<String, dynamic>> query = _ticketsCol;
    if (status != null) {
      query = query.where('status', isEqualTo: status.name);
    }
    query = query.orderBy('createdAt', descending: true);
    if (startAfter != null) query = query.startAfterDocument(startAfter);
    final snap = await query.limit(limit + 1).get();
    final docs = snap.docs;
    final hasMore = docs.length > limit;
    final pageDocs = hasMore ? docs.sublist(0, limit) : docs;
    return TicketPage(
      tickets: pageDocs.map((d) => Ticket.fromMap(d.data())).toList(),
      lastDoc: pageDocs.isEmpty ? null : pageDocs.last,
      hasMore: hasMore,
    );
  }

  Future<Ticket?> fetchTicket(String id) async {
    final doc = await _ticketsCol.doc(id).get();
    if (!doc.exists || doc.data() == null) return null;
    return Ticket.fromMap(doc.data()!);
  }

  /// The job created for [ticketId], if any — a ticket only has one once a
  /// provider is confirmed.
  Future<Job?> fetchJobForTicket(String ticketId) async {
    final snap = await _jobsCol
        .where('ticketId', isEqualTo: ticketId)
        .limit(1)
        .get();
    if (snap.docs.isEmpty) return null;
    return Job.fromMap(snap.docs.first.data());
  }

  /// Cancels a ticket that is stuck — e.g. flagged in a report and needs to
  /// stop broadcasting immediately. Mirrors what
  /// `TicketService._syncBroadcast` does for a normal status change: the
  /// redacted `openTickets` projection is removed in the same batch, so no
  /// provider keeps seeing a request an admin has just killed.
  Future<void> forceCancelTicket(Ticket ticket, {required String reason}) async {
    final batch = FirebaseService.instance.firestore.batch();
    batch.update(_ticketsCol.doc(ticket.id), {
      'status': TicketStatus.cancelled.name,
      'updatedAt': fs.Timestamp.fromDate(DateTime.now()),
    });
    batch.delete(_openTicketsCol.doc(ticket.id));
    await batch.commit();
    await AdminAuditService.instance.log(
      action: 'ticket.force_cancel',
      targetType: 'ticket',
      targetId: ticket.id,
      details: {'reason': reason, 'previousStatus': ticket.status.name},
    );
  }

  Future<JobPage> fetchJobPage({
    JobStatus? status,
    fs.DocumentSnapshot<Map<String, dynamic>>? startAfter,
    int limit = 25,
  }) async {
    fs.Query<Map<String, dynamic>> query = _jobsCol;
    if (status != null) {
      query = query.where('status', isEqualTo: status.name);
    }
    query = query.orderBy('createdAt', descending: true);
    if (startAfter != null) query = query.startAfterDocument(startAfter);
    final snap = await query.limit(limit + 1).get();
    final docs = snap.docs;
    final hasMore = docs.length > limit;
    final pageDocs = hasMore ? docs.sublist(0, limit) : docs;
    return JobPage(
      jobs: pageDocs.map((d) => Job.fromMap(d.data())).toList(),
      lastDoc: pageDocs.isEmpty ? null : pageDocs.last,
      hasMore: hasMore,
    );
  }

  Future<Job?> fetchJob(String id) async {
    final doc = await _jobsCol.doc(id).get();
    if (!doc.exists || doc.data() == null) return null;
    return Job.fromMap(doc.data()!);
  }

  Future<void> forceCancelJob(Job job, {required String reason}) async {
    await _jobsCol.doc(job.jobId).update({
      'status': JobStatus.cancelled.name,
      'cancellationReason': reason,
    });
    await AdminAuditService.instance.log(
      action: 'job.force_cancel',
      targetType: 'job',
      targetId: job.jobId,
      details: {'reason': reason, 'previousStatus': job.status.name},
    );
  }
}

File: admin-web/src/services/adminTicketJobService.ts
import {
  collection, doc, getDoc, getDocs, writeBatch, updateDoc,
  query, where, orderBy, startAfter, limit as fsLimit, Timestamp,
  type QueryDocumentSnapshot, type DocumentData,
} from 'firebase/firestore'
import { db } from '@/lib/firebase'
import { logAuditEvent } from './adminAuditService'
import { mapTicket, mapJob, type Ticket, type Job } from '@/types/models'
import type { TicketStatus, JobStatus } from '@/types/enums'

const ticketsCol = () => collection(db, 'tickets')
const openTicketsCol = () => collection(db, 'openTickets')
const jobsCol = () => collection(db, 'jobs')

export interface TicketPage {
  tickets: Ticket[]
  lastDoc: QueryDocumentSnapshot<DocumentData> | null
  hasMore: boolean
}

export interface JobPage {
  jobs: Job[]
  lastDoc: QueryDocumentSnapshot<DocumentData> | null
  hasMore: boolean
}

export async function fetchTicketPage(opts: {
  status?: TicketStatus
  startAfterDoc?: QueryDocumentSnapshot<DocumentData>
  pageLimit?: number
}): Promise<TicketPage> {
  const limitValue = opts.pageLimit ?? 25
  const constraints = []
  if (opts.status) constraints.push(where('status', '==', opts.status))
  constraints.push(orderBy('createdAt', 'desc'))
  if (opts.startAfterDoc) constraints.push(startAfter(opts.startAfterDoc))
  constraints.push(fsLimit(limitValue + 1))

  const snap = await getDocs(query(ticketsCol(), ...constraints))
  const docs = snap.docs
  const hasMore = docs.length > limitValue
  const pageDocs = hasMore ? docs.slice(0, limitValue) : docs
  return {
    tickets: pageDocs.map((d) => mapTicket(d.id, d.data())),
    lastDoc: pageDocs.length ? pageDocs[pageDocs.length - 1] : null,
    hasMore,
  }
}

export async function fetchTicket(id: string): Promise<Ticket | null> {
  const snap = await getDoc(doc(ticketsCol(), id))
  return snap.exists() ? mapTicket(snap.id, snap.data()) : null
}

/** The job created for ticketId, if any — a ticket only has one once a provider is confirmed. */
export async function fetchJobForTicket(ticketId: string): Promise<Job | null> {
  const snap = await getDocs(query(jobsCol(), where('ticketId', '==', ticketId), fsLimit(1)))
  return snap.empty ? null : mapJob(snap.docs[0]!.id, snap.docs[0]!.data())
}

/** Cancels a stuck ticket, removing its openTickets broadcast projection in the same batch. */
export async function forceCancelTicket(ticket: Ticket, reason: string): Promise<void> {
  const batch = writeBatch(db)
  batch.update(doc(ticketsCol(), ticket.id), {
    status: 'cancelled' satisfies TicketStatus,
    updatedAt: Timestamp.fromDate(new Date()),
  })
  batch.delete(doc(openTicketsCol(), ticket.id))
  await batch.commit()
  await logAuditEvent({
    action: 'ticket.force_cancel',
    targetType: 'ticket',
    targetId: ticket.id,
    details: { reason, previousStatus: ticket.status },
  })
}

export async function fetchJobPage(opts: {
  status?: JobStatus
  startAfterDoc?: QueryDocumentSnapshot<DocumentData>
  pageLimit?: number
}): Promise<JobPage> {
  const limitValue = opts.pageLimit ?? 25
  const constraints = []
  if (opts.status) constraints.push(where('status', '==', opts.status))
  constraints.push(orderBy('createdAt', 'desc'))
  if (opts.startAfterDoc) constraints.push(startAfter(opts.startAfterDoc))
  constraints.push(fsLimit(limitValue + 1))

  const snap = await getDocs(query(jobsCol(), ...constraints))
  const docs = snap.docs
  const hasMore = docs.length > limitValue
  const pageDocs = hasMore ? docs.slice(0, limitValue) : docs
  return {
    jobs: pageDocs.map((d) => mapJob(d.id, d.data())),
    lastDoc: pageDocs.length ? pageDocs[pageDocs.length - 1] : null,
    hasMore,
  }
}

export async function fetchJob(id: string): Promise<Job | null> {
  const snap = await getDoc(doc(jobsCol(), id))
  return snap.exists() ? mapJob(snap.id, snap.data()) : null
}

export async function forceCancelJob(job: Job, reason: string): Promise<void> {
  await updateDoc(doc(jobsCol(), job.jobId), {
    status: 'cancelled' satisfies JobStatus,
    cancellationReason: reason,
  })
  await logAuditEvent({
    action: 'job.force_cancel',
    targetType: 'job',
    targetId: job.jobId,
    details: { reason, previousStatus: job.status },
  })
}
